platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/internal/services/ToolchainsJvmServices.java

```diff
@@ -0,0 +1,94 @@
+import org.gradle.internal.jvm.inspection.JvmMetadataDetector;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.os.OperatingSystem;
+import org.gradle.internal.reflect.Instantiator;
+import org.gradle.internal.service.ServiceRegistration;
+import org.gradle.internal.service.scopes.AbstractPluginServiceRegistry;
+import org.gradle.jvm.toolchain.JavaToolchainResolverRegistry;
+import org.gradle.jvm.toolchain.internal.AutoInstalledInstallationSupplier;
+import org.gradle.jvm.toolchain.internal.DefaultJavaToolchainResolverRegistry;
+import org.gradle.jvm.toolchain.internal.DefaultJavaToolchainService;
+import org.gradle.jvm.toolchain.internal.DefaultJvmToolchainManagement;
+import org.gradle.jvm.toolchain.internal.InstallationSupplier;
+import org.gradle.jvm.toolchain.internal.JavaToolchainQueryService;
+import org.gradle.jvm.toolchain.internal.install.DefaultJavaToolchainProvisioningService;
+import org.gradle.jvm.toolchain.internal.install.JdkCacheDirectory;
+import org.gradle.jvm.toolchain.internal.install.SecureFileDownloader;
+import org.gradle.platform.internal.DefaultBuildPlatform;
+
+import java.util.List;
+
+public class ToolchainsJvmServices extends AbstractPluginServiceRegistry {
+    protected static class BuildServices {
+
+        protected DefaultBuildPlatform createBuildPlatform(ObjectFactory objectFactory, SystemInfo systemInfo, OperatingSystem operatingSystem) {
+            return objectFactory.newInstance(DefaultBuildPlatform.class, systemInfo, operatingSystem);
+        }
+
+        protected DefaultJavaToolchainResolverRegistry createJavaToolchainResolverRegistry(
+            Gradle gradle,
+            Instantiator instantiator,
+            ObjectFactory objectFactory,
+            ProviderFactory providerFactory,
+            AuthenticationSchemeRegistry authenticationSchemeRegistry) {
+            return objectFactory.newInstance(DefaultJavaToolchainResolverRegistry.class, gradle, instantiator, objectFactory, providerFactory, authenticationSchemeRegistry);
+        }
+
+        protected DefaultJvmToolchainManagement createToolchainManagement(ObjectFactory objectFactory, JavaToolchainResolverRegistry registry) {
+            return objectFactory.newInstance(DefaultJvmToolchainManagement.class, registry);
+        }
+
+        protected JdkCacheDirectory createJdkCacheDirectory(ObjectFactory objectFactory, GradleUserHomeDirProvider homeDirProvider, FileOperations operations, FileLockManager lockManager, JvmMetadataDetector detector) {
+            return objectFactory.newInstance(JdkCacheDirectory.class, homeDirProvider, operations, lockManager, detector);
+        }
+
+        protected JavaInstallationRegistry createJavaInstallationRegistry(ObjectFactory objectFactory, List<InstallationSupplier> suppliers, BuildOperationExecutor executor, OperatingSystem os) {
+            return objectFactory.newInstance(JavaInstallationRegistry.class, suppliers, executor, os);
+        }
```

Comment:
🤔 I think all of these could be replaced by just adding them to the `registerBuildServices` below with `registration.add(...)`

Reply:
This works for everything except for`DefaultJavaToolchainResolverRegistry`, which throws `InstantiationException` and can't be created.